folding-schemes/src/commitment/ipa.rs

```diff
@@ -0,0 +1,693 @@
+    Ok(s)
+}
+fn build_s_gadget<F: PrimeField, CF: PrimeField>(
+    u: &[NonNativeFieldVar<F, CF>],
+    // u_invs are assumed have their correctness checked in higher levels of the circuit logic
+    u_invs: &[NonNativeFieldVar<F, CF>],
+    k: usize,
+) -> Result<Vec<NonNativeFieldVar<F, CF>>, SynthesisError> {
+    let d: usize = 2_u64.pow(k as u32) as usize;
+    let mut s: Vec<NonNativeFieldVar<F, CF>> = vec![NonNativeFieldVar::one(); d];
+    for i in 0..d / 2 {
+        let i_bits = bit_decompose(i as u64, k);
+        for j in 0..k {
+            if i_bits[j] {
+                s[i] *= u[j].clone();
+            } else {
+                s[i] *= u_invs[j].clone();
+            }
+        }
+
+        // now place the inverse to the other side
+        s[d - 1 - i] = s[i].inverse()?;
+    }
+    Ok(s)
+}
+
+fn inner_prod<F: PrimeField>(a: &[F], b: &[F]) -> Result<F, Error> {
+    if a.len() != b.len() {
+        return Err(Error::NotSameLength(
+            "a".to_string(),
+            a.len(),
+            "b".to_string(),
+            b.len(),
+        ));
+    }
+    let mut c: F = F::zero();
+    for i in 0..a.len() {
+        c += a[i] * b[i];
+    }
+    Ok(c)
+}
+
+// g(x, u_1, u_2, ..., u_k) = <s, b>, naively takes linear, but can compute in log time through
+// g(x, u_1, u_2, ..., u_k) = (\Prod u_i x^{2^i} + u_i^-1) * x
+fn s_b_inner<F: PrimeField>(u: &[F], x: &F) -> Result<F, Error> {
+    let mut c: F = F::one();
+    let mut x_2_i = *x; // x_2_i is x^{2^i}, starting from x^{2^0}=x
+    for u_i in u.iter() {
+        c *= (*u_i * x_2_i)
+            + u_i
+                .inverse()
+                .ok_or(Error::Other("error on computing inverse".to_string()))?;
+        x_2_i *= x_2_i;
+    }
+    Ok(c * x)
+}
+
+// g(x, u_1, u_2, ..., u_k) = <s, b>, naively takes linear, but can compute in log time through
+// g(x, u_1, u_2, ..., u_k) = (\Prod u_i x^{2^i} + u_i^-1) * x
+fn s_b_inner_gadget<F: PrimeField, CF: PrimeField>(
+    u: &[NonNativeFieldVar<F, CF>],
+    x: &NonNativeFieldVar<F, CF>,
+) -> Result<NonNativeFieldVar<F, CF>, SynthesisError> {
+    let mut c: NonNativeFieldVar<F, CF> = NonNativeFieldVar::<F, CF>::one();
+    let mut x_2_i = x.clone(); // x_2_i is x^{2^i}, starting from x^{2^0}=x
+    for u_i in u.iter() {
+        c *= u_i.clone() * x_2_i.clone() + u_i.inverse()?;
+        x_2_i *= x_2_i.clone();
+    }
+    Ok(c * x.clone())
+}
+
+fn powers_of<F: PrimeField>(x: F, n: usize) -> Vec<F> {
+    let mut c: Vec<F> = vec![F::zero(); n];
+    c[0] = x;
+    for i in 1..n {
+        c[i] = c[i - 1] * x;
+    }
+    c
+}
+
+pub type CF<C> = <<C as CurveGroup>::BaseField as Field>::BasePrimeField;
+
+pub struct ProofVar<C: CurveGroup, GC: CurveVar<C, CF<C>>> {
+    a: NonNativeFieldVar<C::ScalarField, CF<C>>,
+    l: Vec<NonNativeFieldVar<C::ScalarField, CF<C>>>,
+    r: Vec<NonNativeFieldVar<C::ScalarField, CF<C>>>,
+    L: Vec<GC>,
+    R: Vec<GC>,
+    u_invs: Vec<NonNativeFieldVar<C::ScalarField, CF<C>>>,
```

Comment:
I feel this is redundant cause for real verifier (not delegated as circuit) the MSM cost should be far more expensive than doing the inverse of `u`, verifier should be okay to compute them itself.

In circuit we could also call `inverse()` and it'd does the same check for us.

Reply:
You're absolutely right!
I ended up in this conclusion for the $u_i^{-2}$, which at some point I tried also to pass them as inputs and just check with the already computed $u_i^2$ the inputted $u_i^{-2}$ by $u_i^2 \cdot u_i^{-2} == 1$, but this (including having the $u_i^{-2}$ as inputs) was taking more constraints than just computing the square of $u_i^2$ directly in-circuit. I've tried now and for the case of $u_i^{-1}$ takes almost the same amount of constraints, and reducing the amount of data in the proof.

Updated it, thanks a lot for spotting this!

Reply:
I see I see